File: web/src/lib/taste-engine.ts
import { FEATURE_KEYS, type AudioFeatures, type FeatureKey } from "./reccobeats";

export interface TasteVector {
  mean: Partial<Record<FeatureKey, number>>;
  std: Partial<Record<FeatureKey, number>>;
  minVal: Partial<Record<FeatureKey, number>>;
  maxVal: Partial<Record<FeatureKey, number>>;
  sampleCount: number;
}

/**
 * Build a taste vector from a map of trackId -> AudioFeatures.
 * Calculates mean, std, min, max for each feature dimension.
 */
export function buildTasteVector(featuresByTrack: Map<string, AudioFeatures>): TasteVector {
  const tv: TasteVector = {
    mean: {},
    std: {},
    minVal: {},
    maxVal: {},
    sampleCount: featuresByTrack.size,
  };

  if (featuresByTrack.size === 0) return tv;

  // Collect all values per feature
  const featureValues: Record<FeatureKey, number[]> = {} as Record<FeatureKey, number[]>;
  for (const key of FEATURE_KEYS) {
    featureValues[key] = [];
  }

  for (const features of featuresByTrack.values()) {
    for (const key of FEATURE_KEYS) {
      if (features[key] != null) {
        featureValues[key].push(features[key]);
      }
    }
  }

  for (const key of FEATURE_KEYS) {
    const vals = featureValues[key];
    if (vals.length === 0) continue;

    const n = vals.length;
    const mean = vals.reduce((sum, v) => sum + v, 0) / n;
    const variance = n > 1
      ? vals.reduce((sum, v) => sum + (v - mean) ** 2, 0) / n
      : 0;

    tv.mean[key] = mean;
    tv.std[key] = Math.sqrt(variance);
    tv.minVal[key] = Math.min(...vals);
    tv.maxVal[key] = Math.max(...vals);
  }

  return tv;
}

/**
 * Normalize features to roughly 0-1 scale.
 * Loudness: typically -60 to 0 dB -> (v + 60) / 60
 * Tempo: typically 60-200 BPM -> (v - 60) / 140
 * All other features: already 0-1
 */
function normalize(features: Partial<Record<FeatureKey, number>>): Partial<Record<FeatureKey, number>> {
  const result: Partial<Record<FeatureKey, number>> = {};
  for (const [k, v] of Object.entries(features) as Array<[FeatureKey, number]>) {
    if (v == null) continue;
    if (k === "loudness") {
      result[k] = (v + 60) / 60;
    } else if (k === "tempo") {
      result[k] = (v - 60) / 140;
    } else {
      result[k] = v;
    }
  }
  return result;
}

/**
 * Compute cosine similarity between two feature dicts.
 * Both are normalized before comparison.
 * Returns value in [-1, 1]; higher = more similar.
 */
export function cosineSimilarity(
  vecA: Partial<Record<FeatureKey, number>>,
  vecB: Partial<Record<FeatureKey, number>>,
): number {
  const normA = normalize(vecA);
  const normB = normalize(vecB);

  const commonKeys = FEATURE_KEYS.filter(
    (k) => normA[k] != null && normB[k] != null,
  );
  if (commonKeys.length === 0) return 0;

  const dot = commonKeys.reduce((sum, k) => sum + normA[k]! * normB[k]!, 0);
  const magA = Math.sqrt(commonKeys.reduce((sum, k) => sum + normA[k]! ** 2, 0));
  const magB = Math.sqrt(commonKeys.reduce((sum, k) => sum + normB[k]! ** 2, 0));

  if (magA === 0 || magB === 0) return 0;
  return dot / (magA * magB);
}

/**
 * Score a candidate track against the taste vector.
 * Combined score: 70% cosine similarity against mean + 30% range-fit (features within 1.5 std).
 */
export function scoreCandidate(
  candidateFeatures: AudioFeatures,
  taste: TasteVector,
): number {
  const similarity = cosineSimilarity(candidateFeatures, taste.mean);

  let withinRange = 0;
  let totalFeatures = 0;

  for (const key of FEATURE_KEYS) {
    const mean = taste.mean[key];
    const std = taste.std[key];
    const val = candidateFeatures[key];
    if (val != null && mean != null && std != null) {
      totalFeatures += 1;
      if (Math.abs(val - mean) <= std * 1.5) {
        withinRange += 1;
      }
    }
  }

  const rangeScore = totalFeatures > 0 ? withinRange / totalFeatures : 0;

  // 70% similarity + 30% range fit (matches Python implementation)
  return 0.7 * similarity + 0.3 * rangeScore;
}

File: src/taste_engine.py
"""Taste engine — builds feature profiles and scores candidates via cosine similarity."""

import math
from dataclasses import dataclass, field

from .reccobeats import FEATURE_KEYS


@dataclass
class TasteVector:
    """Mathematical profile of what the user likes, built from audio features."""

    mean: dict[str, float] = field(default_factory=dict)
    std: dict[str, float] = field(default_factory=dict)
    min_val: dict[str, float] = field(default_factory=dict)
    max_val: dict[str, float] = field(default_factory=dict)
    sample_count: int = 0


def build_taste_vector(features_by_track: dict[str, dict[str, float]]) -> TasteVector:
    """Build a taste vector from a collection of track features.

    Calculates mean, std, min, max for each feature dimension.
    """
    if not features_by_track:
        return TasteVector()

    # Collect all values per feature
    feature_values: dict[str, list[float]] = {k: [] for k in FEATURE_KEYS}
    for features in features_by_track.values():
        for k in FEATURE_KEYS:
            if k in features:
                feature_values[k].append(features[k])

    tv = TasteVector(sample_count=len(features_by_track))

    for k in FEATURE_KEYS:
        vals = feature_values[k]
        if not vals:
            continue
        n = len(vals)
        mean = sum(vals) / n
        variance = sum((v - mean) ** 2 for v in vals) / n if n > 1 else 0
        tv.mean[k] = mean
        tv.std[k] = math.sqrt(variance)
        tv.min_val[k] = min(vals)
        tv.max_val[k] = max(vals)

    return tv


def cosine_similarity(vec_a: dict[str, float], vec_b: dict[str, float]) -> float:
    """Compute cosine similarity between two feature dicts.

    Normalizes features to 0-1 range before computing similarity.
    Returns value between -1 and 1 (higher = more similar).
    """
    # Normalize loudness and tempo to 0-1 range
    normalized_a = _normalize(vec_a)
    normalized_b = _normalize(vec_b)

    common_keys = set(normalized_a.keys()) & set(normalized_b.keys())
    if not common_keys:
        return 0.0

    dot = sum(normalized_a[k] * normalized_b[k] for k in common_keys)
    mag_a = math.sqrt(sum(normalized_a[k] ** 2 for k in common_keys))
    mag_b = math.sqrt(sum(normalized_b[k] ** 2 for k in common_keys))

    if mag_a == 0 or mag_b == 0:
        return 0.0

    return dot / (mag_a * mag_b)


def _normalize(features: dict[str, float]) -> dict[str, float]:
    """Normalize features to roughly 0-1 scale."""
    result = {}
    for k, v in features.items():
        if k == "loudness":
            # Loudness typically -60 to 0 dB
            result[k] = (v + 60) / 60
        elif k == "tempo":
            # Tempo typically 60-200 BPM
            result[k] = (v - 60) / 140
        elif k == "key":
            # Key 0-11
            result[k] = v / 11
        elif k == "mode":
            # Mode 0 or 1
            result[k] = float(v)
        else:
            # Other features already 0-1
            result[k] = v
    return result


def score_candidate(
    candidate_features: dict[str, float],
    taste: TasteVector,
) -> float:
    """Score a candidate track against the taste vector.

    Uses cosine similarity against the mean vector,
    with a penalty for features outside the taste range.
    """
    similarity = cosine_similarity(candidate_features, taste.mean)

    # Bonus/penalty based on how many features fall within 1 std of the mean
    within_range = 0
    total_features = 0
    for k in FEATURE_KEYS:
        if k in candidate_features and k in taste.mean and k in taste.std:
            total_features += 1
            diff = abs(candidate_features[k] - taste.mean[k])
            if diff <= taste.std[k] * 1.5:
                within_range += 1

    range_score = within_range / total_features if total_features > 0 else 0

    # Combined score: 70% similarity, 30% range fit
    return 0.7 * similarity + 0.3 * range_score
